### ISSUE TITLE
Very high CPU usage on MacOS

### ISSUE THREAD

**Issue by hmans** [bevyengine/bevy#10261]
## Summary

On MacOS, even simple Bevy apps cause very high CPU load; for example, even a blank app that just loads `DefaultPlugins` to render a blank window sits at ~50% CPU usage on an Apple Silicon M1. Add a couple of rendered meshes and postprocessing and it'll quickly go up to about ~90% CPU.

On Windows and Linux, the same app will comfortable sit below 1% (even with some light rendering.)

Additional notes:

- [A MRP has been provided here](https://github.com/hmans/bevy-mrp-macos-performance).
- It uses the `opt-level = 3` etc. trick from the docs.
- Compiling a `--release` build doesn't make a difference.
- In addition, when hiding the app, CPU load doubles, but I understand there have already been issues filed about this particular problem.

## Bevy version

`0.11.3` (but I also tried `66f72dd`, the newest commit on `main` at the time.)

## Relevant system information

- Rust 1.73.0
- MacOS Sonoma 14.0

```
AdapterInfo { name: "Apple M1", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
SystemInfo { os: "MacOS 14.0 ", kernel: "23.0.0", cpu: "Apple M1", core_count: "8", memory: "16.0 GiB" }
```

## What you did

- Build a "blank" Bevy app ([code available here](https://github.com/hmans/bevy-mrp-macos-performance))
- Compile and run it

## What went wrong

Expected behavior: the app should cause CPU load in the single-digit percentage range, like it does on Windows (<1% CPU, on a Ryzen 7.)

Actual behavior: the app caused ~50% CPU load.

**Comment by alice-i-cecile** [comment]
This is a long-standing bug, but it's nice to see it seemingly isolated to Mac. Could we see a tracing log? https://github.com/bevyengine/bevy/blob/main/docs/profiling.md

I'd also like to try stripping out even more things from the `DefaultPlugins`, and see if we can reduce this to one or two offending plugins. Unfortunately I don't have access to a Mac to test this myself, sorry.

**Comment by hmans** [comment]
I've created a tracing log using `cargo run --release --features bevy/trace_chrome`, letting run the app run for a handful of seconds, but it's already 442 MB, so I can't upload it here. :(

Is there any way I can send it to you, via Perfetto or otherwise?

**Comment by hmans** [comment]
Assuming that this represents a (randomly picked) single frame, maybe this is of use to you:

<img width="1083" alt="image" src="https://github.com/bevyengine/bevy/assets/1061/7859537e-3036-4f40-bbb7-6bdaf69d17e7">

**Comment by alice-i-cecile** [comment]
Maybe a GDrive link? Feel free to delete it after a couple days.

@superdump it looks like we're getting hit by excessively long prepare systems again 🤔

**Comment by hmans** [comment]
@alice-i-cecile - here's a Google Drive link for the trace, I hope it works for you:

https://drive.google.com/file/d/1DxPcWasjVDRJDlL1_PSQh-5qhpxEeA3p/view?usp=sharing

**Comment by alice-i-cecile** [comment]
Trace downloaded and open locally, thank you very much!

Pulling this up in Perfetto, I'm still just seeing extremely long Prepare systems. I wonder if the way that wgpu polls is different on Mac? Perhaps it's spinning a thread to try and vsync automatically?

**Comment by hmans** [comment]
> Pulling this up in Perfetto, I'm still just seeing extremely long Prepare systems. I wonder if the way that wgpu polls is different on Mac? Perhaps it's spinning a thread to try and vsync automatically?

Not sure how helpful this is, but I did try a couple of the more simple `wgpu` examples (from their latest `main`) and didn't observe any noticeable performance issues that looked like the one I was seeing with Bevy.

But please keep in mind that I'm new to this entire ecosystem (including Rust itself), so YMMV...

**Comment by alice-i-cecile** [comment]
That's very helpful feedback.

This might be related to https://github.com/bevyengine/bevy/issues/9964.

**Comment by superdump** [comment]
Reduce frame rate to what?

**Comment by Vrixyz** [comment]
I reduced the amount of code necessary to reproduce out on https://github.com/hmans/bevy-mrp-macos-performance/pull/1

**Comment by Vrixyz** [comment]
Seems related to https://github.com/bevyengine/bevy/issues/5713

**Comment by matevz-ap** [comment]
I noticed that my MacBook was running hot, and I found this issue. Upon running the [example provided](https://bevyengine.org/learn/book/getting-started/) with `DefaultPlugins`, I noticed an unusually high CPU usage of 170%. Through a process of elimination, I determined that the culprit behind the issue is the `RenderPlugin`. The example runs smoothly with normal CPU usage when this particular plugin is excluded. While I intend to delve further into this matter, my limited experience with Bevy, having used it for only a day, may limit the extent of my assistance.

**Comment by dmlary** [comment]
So I dug into this about 5 months ago, and found the same problem somewhere in the render pipeline. The discussion & profiling results are in the discord thread below. 

It looked like it may have been a contention issue in wgpu.  That was about where I stopped because needed to make a wgpu-only poc to see if it was a wgpu-on-mac issue. Also this research was done on an Intel Mac. 

https://discord.com/channels/691052431525675048/743663924229963868/1119076766405886004

**Comment by hmans** [comment]
I've just checked the current `latest` and this still seems to be an issue (2 full cores used on a M2 Pro with even the trivial examples.) Is there any outlook on this eventually improving?

**Comment by alice-i-cecile** [comment]
We are interested in this improving! But this needs further investigation to pinpoint: it's still not at all clear where in the stack this is occurring. Does this occur with `wgpu`'s examples?

**Comment by adals** [comment]
[related issue in my opinion ](https://github.com/bevyengine/bevy/issues/13314)

**Comment by hmans** [comment]
> We are interested in this improving! But this needs further investigation to pinpoint: it's still not at all clear where in the stack this is occurring. Does this occur with `wgpu`'s examples?

I just tried exactly this today (current `HEAD` there) and it seems fine. I'm happy to provide further testing or any other input that might help. Sadly I am not familiar with either codebase so I can't really provide any input there.

**Comment by alice-i-cecile** [comment]
Great :) I would double-check at https://github.com/gfx-rs/wgpu/releases/tag/v0.19.4 which is the version Bevy is currently on, but I would suspect that there seems to be something wrong with Bevy's implementation then. 

It looks like the various Prepare steps are at fault. I'll ask the rendering experts if they might have insight there. I suspect we're spin locking while waiting to coordinate between the various steps somewhere, which is manifesting as high CPU usage without actually doing useful work.

Are you able to achieve reasonable results (more than 50k entities) on the Bevymark example?

**Comment by IceSentry** [comment]
@hmans could you try forcing a scale_factor of 1.0 for the window and see if it makes a difference. For some reason macos reports a really high scale factor by default.

**Comment by superdump** [comment]
I tested with 0.11.3 and 0.14.0-rc.2 and observe the same. Using scale_factor of 1.0 makes no difference.

I looked at a GPU frame trace in Xcode and it's only running the no_camera_clear_pass and using microseconds of time. Oddly it said that pass was using 10800 vertices. No idea why.

I looked at a CPU profile using Instruments and that shows no significant CPU usage. Running for 20s put CPU cycles spent for frame updates (at 120Hz) at about the same as plugin initialisation.

[2024-06-08 - bevy macos default plugins cpu.trace.zip](https://github.com/user-attachments/files/15746482/2024-06-08.-.bevy.macos.default.plugins.cpu.trace.zip)

**Comment by superdump** [comment]
[bevy-mrp-macos-performance.gputrace.zip](https://github.com/user-attachments/files/15746501/bevy-mrp-macos-performance.gputrace.zip)
Here's the GPU trace with scale factor 1.0.

**Comment by superdump** [comment]
And just to note that I think tracy traces showing waiting on prepare set systems makes sense as they are blocked by waiting on vsync for a new swapchain texture.

**Comment by hmans** [comment]
I've tried 0.14 and the issue is still persisting.

Since wgpu itself does not seem to have this issue, from the perspective of a naive outsider who's not familiar with either code bases, my best guess would be that wgpu uses vsync out of the box (idling until it's time to render the next frame) and Bevy doesn't (at least on macOS); however, **even the "fps_overlay" example that renders literally nothing** uses ~100% CPU while displaying an FPS count of 100 (on my 100Hz screen) when focused, and ~60 FPS when unfocused.

A couple of folks on Reddit pointed me towards the (new?) `WinitSettings::desktop_app()`, but that isn't the solution to this issue because all it does is prevent the rendering of new frames if there is no user input. Once it does render frames, things are as described above.

**Comment by hmans** [comment]
> Great :) I would double-check at https://github.com/gfx-rs/wgpu/releases/tag/v0.19.4 which is the version Bevy is currently on, but I would suspect that there seems to be something wrong with Bevy's implementation then.

I know I've taken my good time with doing this (apologies), but I now got around to checking that specific version, and its `cube` example uses a relaxed ~19% CPU time for rendering at 120 FPS (I'm on a M2 Pro now.)

Bevy 0.14 appears to use wgpu 0.20, which exhibits the same performance.

19% still feel a little too high for what the example is doing, but it's nowhere near the 50%-200% that I'm seeing with simple Bevy examples.

As mentioned on Reddit earlier today, I'm (stupidly and naively, and very likely unsuccessfully) trying to track this issue down myself now. Just wanted to post these numbers here for reference.

**Comment by katopz** [comment]
I can reproduce this by run Sprite Sheet example on m3max, it took 66-75% CPU and 34% CPU for `opt-level = 1`
<img width="344" alt="image" src="https://github.com/user-attachments/assets/dfeb9095-695d-40e5-886b-e5cbbf8946d6">

Not looking good for just 1 sprite animation.

**Comment by IceSentry** [comment]
Can you try with `opt-level=3` or just `--release`? `opt-level=1` is barely any optimization and isn't a good indicator of real performance.

**Comment by katopz** [comment]
> Can you try with `opt-level=3` or just `--release`? `opt-level=1` is barely any optimization and isn't a good indicator of real performance.

`opt-level=1` 👉 34% CPU
`opt-level=3` 👉 36% CPU

i didn't bother `--release` because i on dev and my m3 fans are spinning 🔥

```
[profile.dev]
opt-level = 3

[profile.release]
lto = true
opt-level = 3
codegen-units = 1
incremental = false
debug = false
```
And the result with `[profile.release]`  is
`--release` 👉 34% CPU

just `--release` 👉 36% CPU

Actually it look same all over setting. Something really weird here.

**Comment by destravous** [comment]
What are the chances this issue is related to how bevy is calling winit? 
On the second to last known regression mentioned here: https://github.com/bevyengine/bevy/issues/11052

Bevy's entire update loop is currently running in the `about_to_wait()` wininit callback instead of `WindowEvent::RedrawRequested`. And there seems to be a whole bunch of workarounds + bevy ecs events passed around. (`about_to_wait()` is not supposed to be used for this, `WindowEvent::RedrawRequested` is supposed to be used instead according to the wininit docs).

See `ApplicationHandler` in `bevy_winit::state.rs`

**Comment by Zajozor** [comment]
> If anyone can provide a recent MPC example of this problem I would gladly help pinpoint the exact issue.

If this still helps - I'm here with a M1 chip (Macbook Pro 2020 M1 16GB RAM).

1. `git clone git@github.com:bevyengine/bevy.git`
2. `cargo run --features="bevy_dev_tools" --example fps_overlay` runs at 60FPS & ~80% CPU (interestingly, when the window is not focused, it jumps to 85-90% CPU fairly consistently)
3. `cargo run --release --features="bevy_dev_tools" --example fps_overlay` runs at 60FPS & ~45% CPU

One interesting thing I noticed is that the CPU jumps considerably when you start moving the mouse over the window while it's not focused. Here's a short example:

https://github.com/user-attachments/assets/e4813bda-1038-459a-b81a-4888ff3158ab

(not sure if it's relevant, just thought it might be)

lmk if I can help reproducing anything, or trying out a fix 🙏🏻

**Comment by wentao** [comment]
I ran into similar issue on iOS (not Mac though). A simple example like https://bevyengine.org/examples/3d-rendering/3d-scene/ running on iOS (with bevy 0.14.2) has 100%+ CPU usage. I have not tried the --release flag yet.

**Comment by voneiden** [comment]
I was looking at a similar issue on Linux/Wayland and figured that [UpdateMode](https://docs.rs/bevy/latest/bevy/winit/enum.UpdateMode.html) makes a really big difference in my case. With `Continuous` (default) I'm getting something like 60-70% CPU usage and with `Reactive` I'm getting about 2%. 

Just throwing it out here, maybe it's of some use to you folks. Considering this issue doesn't seem to affect everyone I'm guessing that `Continuous` is not a CPU hog for most users. Perhaps some OS/WM level thing that limits redraw frequency? https://github.com/rust-windowing/winit/issues/3933 while not related to bevy, gives me the impression that in some cases `request_redraw` (which bevy uses) might end up drawing the window at a really high frequency - way higher than what bevy is rendering at. In the linked issue the reporter observed 76000 redraws per second.

**Comment by hacknus** [comment]
I also experience this problem on my MacBook Pro M2 Pro with macOS 15.3.2


with
```rust
        .insert_resource(WinitSettings::desktop_app())
```
the CPU usage drops to zero during idle, but it spikes up to 50% - 70% when moving the cursor.

@Kees-van-Beilen Have you made any further progress in detecting the issue?

**Comment by hacknus** [comment]
![Image](https://github.com/user-attachments/assets/50f5e6ac-cbc9-486c-baf2-3e42cf37b521)

This is a cargo flamegraph profiling output of
```shell
 cargo flamegraph --features="bevy_dev_tools" --example fps_overlay
```

It sits at 70% CPU usage.

**Comment by mockersf** [comment]
Higher cpu usage on debug profile is expected.

If you compare with Windows, macOS reports CPU usage differently: on Windows the percentage is across all cores, on macOS it's per cores. With a M2 Pro you have 10 or 12 cores, so to compare with Windows you would have to divide your 70% by 10 or 12, so 6~7%

**Comment by hacknus** [comment]
> Higher cpu usage on debug profile is expected.
> 
> If you compare with Windows, macOS reports CPU usage differently: on Windows the percentage is across all cores, on macOS it's per cores. With a M2 Pro you have 10 or 12 cores, so to compare with Windows you would have to divide your 70% by 10 or 12, so 6~7%

Thank you for the reply. Cargo Flamegraph automatically builds in release. 
Yes, macOS shows per core, but a simple egui application only shows max 10% of usage even when interacting and moving the cursor. An empty bevy app uses more than double, which I don't see why this should be expected.

When moving an egui app from eframe to bevy_egui, the CPU usage more than doubles.

**Comment by hacknus** [comment]
@Kees-van-Beilen thanks for the insight! I'm looking forward to your investigation!

I would still expect an empty bevy-window to have a comparable CPU usage to other simple applications (so around 10%) even with the per-core reporting of macOS.

**Comment by hacknus** [comment]
I performed some tests by disabling certain features:

code:
```rust
use bevy::{
    prelude::*,

};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins) 
        .run();
}
```

- with minimum features we only use about 14-18% CPU: 
```toml
bevy = { version = "0.16", default-features = false, features = [
    "bevy_render",
    "bevy_window",
    "bevy_winit",
    "bevy_core_pipeline",
    "x11" # or "wayland" depending on your system (on Linux)
] }
```

- by including the `multi_threaded` feature, the cpu usage rises from 20% to 30 - 40%.

However, with my more complex bevy application, I also get 80% CPU without the `multi_threaded` feature

**Comment by hacknus** [comment]
As a reference: on windows the minimum usage example reports 1.1 % usage. Accounted for the per cpu reporting this would be equal to about 9-10% of cpu usage per core, which is still half or at least a third less in usage than on macOS.

**Comment by Vricken** [comment]
An empty scene with just an fps counter uses between 60% to 70% of my CPU and 20% of my GPU. I'm on an m4 pro chip so an empty scene with an fps counter should definitely not take that much.

**Comment by alice-i-cecile** [comment]
@IceSentry [did some preliminary experiments](https://discord.com/channels/691052431525675048/1253771396832821270/1391957084794257581) today with an alternate SDL2 backend, without seeing the large CPU usage drop that we were hoping for.

Similarly, he tried swapping to the single-threaded schedule executor but no improvements there.

That basically just leaves wgpu in some form, in terms of "platform specific code that might be causing terrible performance problems". Unless there's like... a serious platform specific bug in audio or gilrs??

**Comment by hmans** [comment]
I don't know if this is going to be helpful, but when I was playing around with wgpu and winit (latest versions of each) a few months ago, I did observe that winit was defaulting to a model where input events were polled _continuously_ (leading to the same sort of CPU usage as described here), and I was able to fix it by changing it to only do it _once per frame_.

(Unfortunately I don't remember what the required change was, I'm not really doing any Rust things and this was in March or so.)

I have no idea

- if this only occurs in macOS vs. Windows (I didn't try Windows)
- if fixing this in Bevy is as straight-forward as it was in direct wgpu+winit
- or if this is an overall Bevy issue (I still know nothing about its codebase.)

But I am 99.9% sure that this is caused by something continuously polling for input events, and the solution must be along the lines of no longer doing that.

Apologies if I'm only stating the very obvious here, there's not a lot I can add (other than I remember it being easy enough for a complete Rust-idiot like myself to set this up in wgpu+winit. I'm just hoping the Bevy fix will be just as simple.)

**Comment by Kees-van-Beilen** [comment]
@alice-i-cecile 
> That basically just leaves wgpu in some form, in terms of "platform specific code that might be causing terrible performance problems". Unless there's like... a serious platform specific bug in audio or gilrs??

I do think that it is wgpu (or how bevy interacts with wgpu) releated. I don't have proof other than that some metal api calls have been made that shouldn't. I haven't yet had the time yet to run an in depth metal analysis. 

@hmans 
> But I am 99.9% sure that this is caused by something continuously polling for input events, and the solution must be along the lines of no longer doing that.

It's a great theory, however this is only true for debug builds. You cannot poll input events just once per frame. As macos does not keep a buffer of events (see https://developer.apple.com/documentation/appkit/nsapplication/nextevent(matching:until:inmode:dequeue:)?language=objc). I was looking at the Winit implementation of event polling and they start and stop the app, which isn't the bestway to do things but in release mode shouldn't matter. I already investigated the runloop in #5713 and it does not seem that for release builds the way how events and the runloop work really matter.

**Comment by IceSentry** [comment]
To add a bit more details on my experiments from last night. My SDL backend (https://github.com/IceSentry/bevy_window_sdl2_backend) which is highly inspired by the winit backend resulted in 70-80% cpu usage which is exactly the same as what the winit backend uses for me. The SDL version is _very_ simple so my intuition tells me the issue most be somewhere else. It's highly suspect that both winit and SDL2 would have the same cpu usage.

After that I tried using the single threaded executor on every schedule which helped a bit and now I saw cpu usage in the range of 60-70%. This isn't a macos issue though, we know the multi threaded executor has overhead and this would reflect on all platform, not just macos. So while it is good to know that we should make it easier to configure since it would help this isn't the root cause.

Finally, I tried using bevy_framepace and cpu usage went down to 30-40%. This is what made the biggest improvement which seems to indicate that the way we are waiting on frames when cpu work is lower than the frame rate is wrong. I think this is also one more argument in favour of upstreaming bevy_framepace. It's not perfect, but it would this issue and many other issues people are experiencing with bevy.

Finally, I tried the wgpu hello_window example which uses winit and that was at 15% cpu. This indicates that wgpu+winit on it's own should be fine and we do something wrong in bevy that makes it use a lot more cpu than necessary.

All my numbers where captured on a plugged in macbook pro with an M4 Max.

**Comment by Kees-van-Beilen** [comment]
Alright, I couldn't resist, i've dropped my work to take another look at whats going on with the macos/ios performance. And ... I might have found something. I'm not quite sure what the performance impact is (probbaly tiny), but this is definitely not a good practise. 

(roughly) Every frame the render pipeline allocates a new temporary buffer and reinterprets the currently used texturebuffers. This is bad practise. MTLBuffers should persist as long as possible (unless a resource property must be changed). Also whilst the texture reinterpretations aren't  costly at all (because nothing happens with the underlying persistent buffer) they do make a small obj-c allocation every frame for every texture.  I'm saying every frame here, but sometimes there are bursts of allocations.

I don't know how much related this is to cpu usage, but this should definitely be fixed. And might give us some more insight in what else might be going wrong with bevy+wgpu+mac 

Maybe we should setup some automated testing these macOS bas practises, so we don't shoot ourselves in the foot. 
Should I open an issue for the MTLBuffer/MTLTexture::create_view  misuse?


Also because debugging memory on macOS can be annoying here is a trace for the 3d_scene example you can put into instruments:
[https://bevy.inpolen.nl/traces/3d_scene.trace.zip](https://bevy.inpolen.nl/traces/3d_scene.trace.zip)

**Comment by alice-i-cecile** [comment]
Please open a new issue and link this one :) That's a very promising lead.

**Comment by Vricken** [comment]
> Alright, I couldn't resist, i've dropped my work to take another look at whats going on with the macos/ios performance. And ... I might have found something. I'm not quite sure what the performance impact is (probbaly tiny), but this is definitely not a good practise. 
> 
> (roughly) Every frame the render pipeline allocates a new temporary buffer and reinterprets the currently used texturebuffers. This is bad practise. MTLBuffers should persist as long as possible (unless a resource property must be changed). Also whilst the texture reinterpretations aren't  costly at all (because nothing happens with the underlying persistent buffer) they do make a small obj-c allocation every frame for every texture.  I'm saying every frame here, but sometimes there are bursts of allocations.
> 
> I don't know how much related this is to cpu usage, but this should definitely be fixed. And might give us some more insight in what else might be going wrong with bevy+wgpu+mac 
> 
> Maybe we should setup some automated testing these macOS bas practises, so we don't shoot ourselves in the foot. 
> Should I open an issue for the MTLBuffer/MTLTexture::create_view  misuse?
> 
> 
> Also because debugging memory on macOS can be annoying here is a trace for the 3d_scene example you can put into instruments:
> [https://bevy.inpolen.nl/traces/3d_scene.trace.zip](https://bevy.inpolen.nl/traces/3d_scene.trace.zip)
> 
> 
> 

Could you maybe test your theory on a build from ~2 years ago? If it's already present this might be very well what is causing the high cpu usage.

**Comment by Kees-van-Beilen** [comment]
Confirmed bevy v0.11.3 commit `0d23d71` has the same issue same for `cb64a4a` (the current main)

**Comment by alice-i-cecile** [comment]
Can we look back even further? This might be bisectable. I don't have a Mac myself to test on, but this is one of our most important open bugs.

**Comment by Kees-van-Beilen** [comment]
Well I already know where one of the misuses is, inside of the `prepare_lights` function
https://github.com/bevyengine/bevy/blob/cb64a4aa8ad6cc0d6398ca46507320c2d54639b2/crates/bevy_pbr/src/render/light.rs#L1239

Luckily when it comes to the textures for shadow maps etc. the metal allocation trace gives the label. Unfortunately, the buffer that being allocated and disposed of every frame has the label "(wgpu internal) Staging" which does not appear in the bevy project

**Comment by Kees-van-Beilen** [comment]
> Can we look back even further? This might be bisectable. I don't have a Mac myself to test on, but this is one of our most important open bugs.

I'll check the mayor releases of bevy

**Comment by superdump** [comment]
Mmm, we should cache those views for shadow mapping. It was assumed that creating views was cheap, but if it isn’t then they should be cached and invalidated when something changes that needs a new texture and view.

Aside from that, I noted on Discord that it sounds like perhaps a wgpu issue given that the wgpu example that just uses a clear colour uses a few tens of % of a core.

**Comment by Kees-van-Beilen** [comment]
I'll check the wgpu example if the same temp buffer is made. 

> Mmm, we should cache those views for shadow mapping. It was assumed that creating views was cheap, but if it isn’t then they should be cached and invalidated when something changes that needs a new texture and view.

Yes those views should be cached, also they only need to be invalidated if their  properties from the TextureViewDescriptor change not the underlying data.  Also it is cheap to create those views, however it screws with metal debugging, and you make 3 objc method calls and 1 obj c allocation per view, which is completely unnecessary and should be avoided.

**Comment by Kees-van-Beilen** [comment]
I've found something interesting in winit. I don't know if this is applicable to bevy. But winit's main eventloop is designed to use 100% cpu (because in the rendering code most programs including bevy send another RedrawRequest). As it will fill all of its remaining cpu budget with ReDrawRequest events. Simply adding a sleep (thread sleep) to these requests drastically improves performance. 

Sorry for all the typos it's ~~1pm~~ 1am for me :)

**Comment by Kees-van-Beilen** [comment]
I might actually be cooking

<img width="228" height="126" alt="Image" src="https://github.com/user-attachments/assets/eb48b2e8-be6c-44ec-aa98-0d26d7df5259" />

**Comment by Vricken** [comment]
> I've found something interesting in winit. I don't know if this is applicable to bevy. But winit's main eventloop is designed to use 100% cpu (because in the rendering code most programs including bevy send another RedrawRequest). As it will fill all of its remaining cpu budget with ReDrawRequest events. Simply adding a sleep (thread sleep) to these requests drastically improves performance. 
> 
> Sorry for all the typos it's 1pm for me :)

On different languages with different libraries i have experienced very very similar things, so you might be on the right path

**Comment by Kees-van-Beilen** [comment]
Sadly, after a bit more testing. this isn't the cpu bottleneck. Bevy already cleverly optimises for the winit RedrawRequest loop. Though the bevy solution is a bit sketchy it works. The main reason I saw this massive performance gain was because I was only measuring the main runloop and how many times it ran. And I didn't notice that app.update() wasn't being called, whoops. This does however confirm that this is not a winit problem.

**Comment by phire** [comment]
I spent many hours trying to narrow this down in Instruments (Apple's profiler), and I just couldn't find any smoking guns. I'm beginning to doubt this is even a bug. 


With a proper scene and a few objects, I was seeing obvious issues the number of gpu resource allocations per frame.   
Bevy probably shouldn't be using [`Queue::write_buffer_with`](https://docs.rs/wgpu/latest/wgpu/struct.Queue.html#method.write_buffer_with), it's documented to create a new allocation on every single call which gets freed at the end of the frame. Maybe it's fine on other GPUs, but on my M1 Pro, each allocation is a minimum of 128KB and requires a full IPC round-trip to the AGX's management co-processor. Bevy should probably be using `[StagingBelt](https://docs.rs/wgpu/latest/wgpu/util/struct.StagingBelt.html)` instead, which recycles a pool of buffers internally. 

But by the time I stripped my reproduction down to simply loading the default plugins, with most default features disabled (especially multithreading and picking), switched to a release build, collecting enough samples, nothing was sticking out as wrong. It was simply taking ~1.5ms of CPU time to process everything bevy wanted to do per frame, spending maybe 0.5ms blocked on a few GPU driver calls, and then sitting idle for ~8ms (on my 100hz screen). 

----------

The problem is that CPU utilisation simply doesn't make much sense anymore. It made a lot of sense when you had a single CPU core locked at a fixed clock speed. But on a modern advanced mobile processor with insane power management features? It's just misleading. 

My minimal reproduction got CPU utilisation down to about 15% (from the 75% of my simple test scene). That's still very high compared to what I could get on a x86 machine, but you really have to ask "15% of what". 

Because it's not 15% of a core running at full speed. It's 15% of a core running in some kind of extreme power saving mode. 

Annoyingly, apple doesn't seem to provide any profiling tools to see what power mode the CPU or SoC in general is in (only the GPU). But we can use `sudo powermetrics --samplers cpu_power` to monitor the overall power usage (the frequency stats here are somewhat useless and don't seem to tell the whole story. I suspect apple are keeping the CPU cores clocked high, but throttling main memory speeds).

The key is to look at the actual CPU power, and see how much it changes when running bevy. 

On my M1 Pro, running my minimal reproduction with ~15% cpu utilisation, it's only actually using ~100 mW of power. That's absolutely tiny, I wish I could get my x86 laptops down that low.  A proper 100% single-threaded load (like `yes > /dev/null`) results in 3500 mW of power usage. If we assume power usage is linear (it's not), that would mean it's really using more like 3% CPU load, which seems reasonable. 

I'm not sure if this is a bug or not; There are clearly performance issues in the profile. Going back to my original scene, powermetrics is more like 850 mW of CPU power while reporting 75% utilisation. If we assume linear power usage again, we could argue this is more like 25% utilisation,  which still seems high compared to the 9% utilisation on my x86 laptop.

**Comment by hacknus** [comment]
Hello @phire 

Thank you very much for your insights! I see what you are saying, but I still think this is a bug in bevy, because regardless of how CPU usage is reported, a very simple `winit` app (window) shows a lot less reported CPU usage than one created with `bevy`. And also compared to other apps (native) the reported CPU usage of `bevy` is much higher. I would argue that the low power usage of the CPU is just due to the insane efficiency of apple silicon ARM architecture, no?

**Comment by hmans** [comment]
Thanks for putting in all this work to narrow down the cause for this issue. I would like to point out though that no other rendering framework or library (that I've tried, and I've tried many; including using plain wgpu with winit) is exhibiting this behavior on macOS. A consensus that this isn't a bug would be very frustrating.

**Comment by phire** [comment]
> a very simple winit app (window) shows a lot less reported CPU usage than one created with bevy.

Yes, but a large part of that is because it does less work. By default, the bevy renderer plugin will extract all the components from the world into the render world, build scene graphs, optimise them and then discover there is no work to be done, then build command buffers anyway and submit a blank frame to the GPU, vsynced. Does your simple winit app even submit work to the gpu?

Like, I said, I'm unsure.... Well, I'm sure CPU time is a useless metric, but that just means it's very hard to tell if a blank bevy_render project is using too much CPU for the amount of work it's doing or not. 

And then there are secondary questions of: Should blank bevy_render project should be making an effort to do less pointless work than it currently does (on all platforms) and; Is bevy_renderer doing too much work when the scene isn't blank?

I think the next step is to try and make a sample project that does about the same amount of work per frame as bevy_render does (or should do) so we have a comparable target. 

> I would argue that the low power usage of the CPU is just due to the insane efficiency of apple silicon ARM architecture, no?

No, the power usage will go much higher if there is actually work to be done. Some single-threaded workloads can drive over 5w and I think multithreaded workloads can hit 35W of CPU power. 100mW is quite low (though the dozen other apps I had open sitting were totalling 50-80mW as long as I didn't move the mouse... so...).

**Comment by phire** [comment]
> A consensus that this isn't a bug would be very frustrating.

Yeah, there is a reason why I went down a rabbit hole of profiling this. Even if it's not a bug, it just feels wrong and gives a bad impression (to both developers and users of their products). 

If we end up there, I think the correct path forwards is to implement some kind of damaged model to bevy_render so it only creates a frame if something changes, and then enable it that default.

**Comment by hacknus** [comment]
> If we end up there, I think the correct path forwards is to implement some kind of damaged model to bevy_render so it only creates a frame if something changes, and then enable it that default.

I think this already exists with 
```rust
        .insert_resource(WinitSettings::desktop_app())
```
which is defined as

```rust
pub fn desktop_app() -> Self {
        WinitSettings {
            focused_mode: UpdateMode::reactive(Duration::from_secs(5)),
            unfocused_mode: UpdateMode::reactive_low_power(Duration::from_secs(60)),
        }
    }
```

**Comment by hacknus** [comment]
> > a very simple winit app (window) shows a lot less reported CPU usage than one created with bevy.
> 
> Yes, but a large part of that is because it does less work. By default, the bevy renderer plugin will extract all the components from the world into the render world, build scene graphs, optimise them and then discover there is no work to be done, then build command buffers anyway and submit a blank frame to the GPU, vsynced. Does your simple winit app even submit work to the gpu?

To be honest, I have not profiled it, so I can't say for sure. But also other frameworks, like `eframe` with `egui` show a lower usage than `bevy_egui` with `egui` (which goes in the direction of what @hmans mentioned, I think). And there I noticed an increase for the same project when I switched frameworks from `eframe` to `bevy_egui`, so I really have the feeling that there is something eating up more resources.


> Like, I said, I'm unsure.... Well, I'm sure CPU time is a useless metric, but that just means it's very hard to tell if a blank bevy_render project is using too much CPU for the amount of work it's doing or not.

Yeah, I see your point.

> No, the power usage will go much higher if there is actually work to be done. Some single-threaded workloads can drive over 5w and I think multithreaded workloads can hit 35W of CPU power. 100mW is quite low (though the dozen other apps I had open sitting were totalling 50-80mW as long as I didn't move the mouse... so...).

Okay, fair enough. Thanks for the clarification!

**Comment by phire** [comment]
I have some bad news.

I hacked together a [rough test for this](https://github.com/phire/winit_vs_bevy/). It uses winit+wgpu to create a window and then submit blank frames. It's still nowhere near as much CPU work as bevy, but it's using almost the same amount of CPU time, 15% vs 18%. It's close enough that I don't think there is a bug.

Though I did find that disabling bevy's multithreading is pretty essential if you want bevy's reported CPU usage to match raw winit+wgpu. No idea if this has any real-world effect. 

------

> I think this already exists with
> ```rust
>        .insert_resource(WinitSettings::desktop_app())
> ```

Ah yes, that does actually work, lowers CPU usage to 0%, as long as you don't move the mouse. 


 > But also other frameworks, like eframe with egui show a lower usage than bevy_egui with egui

From memory, the default eframe+egui doesn't render anything unless there are input events.

**Comment by hacknus** [comment]
> I have some bad news.
> 
> I hacked together a [rough test for this](https://github.com/phire/winit_vs_bevy/). It uses winit+wgpu to create a window and then submit blank frames. It's still nowhere near as much CPU work as bevy, but it's using almost the same amount of CPU time, 15% vs 18%. It's close enough that I don't think there is a bug.

Hmm okay that is indeed bad news. Do you mean by "nowhere near as much CPU work" that in principle there should be less tasks but it still shows comparable CPU load?
 
> Though I did find that disabling bevy's multithreading is pretty essential if you want bevy's reported CPU usage to match raw winit+wgpu. No idea if this has any real-world effect.
> 

yeah, that's weird...

> > But also other frameworks, like eframe with egui show a lower usage than bevy_egui with egui
> 
> From memory, the default eframe+egui doesn't render anything unless there are input events.

Yes, the way I tested it I continuously updated the entire window, not limited to input events.

I also noticed issues with `eframe` on macOS which I noted here: https://github.com/emilk/egui/issues/7059 there seems to be a performance decrease on macOS after version `0.23.0`... But in my experience the reported CPU usage is still lower than that of `bevy`.

Do you have an idea what a test application should contain that can be compared between `eframe` and `bevy`?

**Comment by Kees-van-Beilen** [comment]
> I have some bad news.
> 
> I hacked together a [rough test for this](https://github.com/phire/winit_vs_bevy/). It uses winit+wgpu to create a window and then submit blank frames. It's still nowhere near as much CPU work as bevy, but it's using almost the same amount of CPU time, 15% vs 18%. It's close enough that I don't think there is a bug.
> 
> Though I did find that disabling bevy's multithreading is pretty essential if you want bevy's reported CPU usage to match raw winit+wgpu. No idea if this has any real-world effect.
> 
> > I think this already exists with
> >        .insert_resource(WinitSettings::desktop_app())
> 
> Ah yes, that does actually work, lowers CPU usage to 0%, as long as you don't move the mouse.
> 
> > But also other frameworks, like eframe with egui show a lower usage than bevy_egui with egui
> 
> From memory, the default eframe+egui doesn't render anything unless there are input events.

I've quickly skimmed  your test and it looks like their is a mistake in the event loop implementation (but i'm not sure; also with mistake I mean code that doesn't run efficiently on macOS). Also thank you for confirming the buffer thing as i wasn't certain which wgpu feature was creating the buffers.

Just checked and in release mode the cpu usage matched my minimal test. And is very reasonable given it is ran at 120 fps. 

I do agree with what has been said about that the 80-200% cpu usage is a bad look for developers and users. And we should find out how to reduce that number. I area to maybe look at is to remove the rendering from bevy and then profile.

**Comment by alice-i-cecile** [comment]
> Though I did find that disabling bevy's multithreading is pretty essential if you want bevy's reported CPU usage to match raw winit+wgpu. No idea if this has any real-world effect.

This is currently unsurprising; earlier profiling shows that our existing multithreading overhead is high. @NthTensor has been investigating this and has been working on a replacement.

**Comment by phire** [comment]
From what I can tell, the roughly 15-20% cpu usage for a completely blank frame is expected. 

I expanded [my rough test](https://github.com/phire/winit_vs_bevy/) with a native appkit+metal backend to avoid both winit and wgpu (plus an eframe backend, as requested by @hacknus), and also cross-checked with an equivalent swift app. See it's readme for more details.

-------

However... Bevy's CPU usage quickly climbs as soon as you do anything more than a completely blank frame. Simply adding a camera object doubles CPU usage (and almost triples power usage) Adding a light and a single triangle will hit reach 42% (though adding more triangles has little effect). Enabling shadows will hit 50% usage.   

This still isn't a massive amount of power, the CPU is still in power saving mode, and we are consuming 200-300mW compared to the 100mW of a blank frame. 

But if I go back to that swift version (which is more or less Xcode's "game" sample), it can display a spinning cube with basically the same CPU usage as the blank frame, just 15% and 100mW of power. 

So this does seem to be a valid bug. It's just that the minimal reproduction requires at least a camera and a mesh. I also doubt it's a Mac-only bug, it's just that the Apple Silicons excellent power efficiency and the weird ways it distorts "cpu usage" as a useful metric makes the bug much more obvious.

**Comment by NthTensor** [comment]
Someone should just open up a few of the demo scenes in dtrace or xcode instruments. It's very likely this is caused by spin-loops either in `bevy_winit` or `bevy_tasks` (possibly both), and the only way to find them will be with fine-grained runtime instrumentation. The output from tracy isn't precise enough, we will need full call-tree timing information.

I'll make a note to do this eventually, but if anyone else would like to have a go I'd be appreciative.

**Comment by IceSentry** [comment]
Also, I mentioned it in one of my comments I think but It's worth trying out those test with bevy_framepace. It made a noticeable difference for me.

**Comment by hacknus** [comment]
> Also, I mentioned it in one of my comments I think but It's worth trying out those test with bevy_framepace. It made a noticeable difference for me.

I can confirm this, but in my opinion this is just masking the issue or a workaround.

**Comment by phire** [comment]
> Also, I mentioned it in one of my comments I think but It's worth trying out those test with bevy_framepace. It made a noticeable difference for me

Adding bevy framepace results in framerate dropping from 100fps to 96fps (on my 100hz monitor). So of course the CPU usage dropped, it's rendering less frames. 

I ran into this kind of foot-gun a few times when writing those tests. I thought something was more efficient, but turns out it was just rendering less frames. 

> Someone should just open up a few of the demo scenes in dtrace or xcode instruments. It's very likely this is caused by spin-loops either in bevy_winit or bevy_tasks (possibly both), and the only way to find them will be with fine-grained runtime instrumentation. 

@NthTensor I've now done extensive profiling in instruments, there just doesn't seem to be anything obviously wrong.

<details>
<summary> main-thread hot functions from high frequency sampling </summary>
<img width="1224" height="848" alt="Image" src="https://github.com/user-attachments/assets/ed50f307-7f62-4003-a643-4ccdb818a40e" />
</details>

No function is particularly hot [^1]. There are no spin locks. Window management stuff doesn't show up at all. 

[^1]: Except for memmove/memset at 2-3% each, which my experience normally indicates memory bottlenecks.   
My theory is that because there is so little actual work, main memory is actually sleeping most of the time. Most of the workload hits cache, but any code that misses cache stalls for quite a long time while memory wakes up. 

I tried many of Instruments tools, but nothing is obviously wrong [^2]. I even went down the rabbit hole wonder if MacOS counted time waiting for objc messages as "running" (it doesn't).

[^2]: Other than the number of resource allocations per frame. But I'm not even sure that's contributing to this problem at all. 

 From what I can tell most of the CPU usage is legitimate work spread out over many bevy systems and gpu drivers they call. 

It's a little easier to see with the Flame Graph:
<img width="3424" height="838" alt="Image" src="https://github.com/user-attachments/assets/466d4f67-f4be-4a1e-992f-db5e81f1f842" />

Part of the problem is that bevy simply does a lot of work.

Adding a single triangle activates an entire PBR pipeline, with a compute shader pre-pass doing culling, light clustering, environment light probes. This is simply not comparable to the simple "hello triangle" that I have in my other backends. 

I'm still not sure if this is a valid bug or not; Bevy is doing more work, so of course CPU usage should be higher...

-------

But usage probably shouldn't be this high.

Bevy (and wgpu) is also doing this work quite inefficiently. Bevy is recreating binding groups every single frame, and wgpu isn't doing binding groups efficiently [^3]; these probably should be cached, or static. The same with texture views. 

[^3]: With wgpu, array bindings result in a buffer allocation. That's one 128KB buffer for every single array binding, and bevy uses several. 

And I previously mentioned buffer. My metal backend can simply write directly into the final gpu buffer, but bevy first extracts that data from the main word, copying it into a vector, then later hands it to wgpu, which allocates a new buffer, copies the data again. Then once all commands are generated, wgpu asks metal to copy it a 3rd time into the final staging.  It's just wasteful [^4]

[^4]: And if I'm right about main memory being a bottleneck, then these four copies of the data, accessed at three different points though the frame, are putting extra pressure on the cache. It could have knock-on effects that slows down all the code down because their data was pushed out of cache. 

The later two copies could be considered a wgpu deficiency, they even have an issue (https://github.com/gfx-rs/wgpu/issues/3698), but bevy should probably be aiming to writing directly to the final GPU buffer during the render extract stage. 


IMO, this issue seems to be more of a "death by a thousand cuts" thing than any single cause. Bevy just isn't very optimised for empty scenes. The overhead isn't that bad, but it shows up really badly on Apple Silicon.

**Comment by duncanfinney** [comment]
Hey everyone! 

I spent a lot of time this weekend fighting some battery usage issues on **iOS** (not MacOS specifically). I was seeing crazy power consumption on iOS - to the point where **the phone was getting hot** with a relatively simple screen. A lot of the cause of that was actually the log spam in the Bevy example

After a bit of playing around with it, I was able to figure it out (kind of -- I think). I just wanted to share some of my findings in case it helps anyone else.

----
**Some of my findings:**
- Something is not working with the way `winit` does the event loop on iPhone
- Ripping out `winit` entirely helped with the battery consumption
- Hooking into CADisplayLink also helped to get things working better
- **Looking at CPU percentage alone is misleading.** 
	- The iPhone has 4 efficiency cores and 2 performance cores. 
	- If you get the render loop tight enough, it will assign the work to efficiency cores and use a lot less power (see the 60hz pic)
	- The percentage looks high overall, but the battery consumption is significantly better
- Performance optimizations that I found impactful:
	- disable `multi_thread` for simple scenes
	- disable `bevy_text` entirely

I found a repository by that shows using `CADisplayLink` with bevy and forked it with my findings: 

- Original: https://github.com/jinleili/bevy-in-app - Thanks @jinleili !
- My fork with some iOS battery hacking stuff: https://github.com/duncanfinney/bevy-in-app
----
**Example test app:**
<img src="https://github.com/duncanfinney/bevy-in-app/blob/main/assets/github/app_screenshots/bouncing_ball.gif?raw=true" alt="Bevy in iOS App" height="400" />

**Power profiles:**
![60hz Fixed](https://github.com/duncanfinney/bevy-in-app/blob/main/assets/github/profiles/60hz_fixed.png?raw=true)
![120hz Fixed](https://github.com/duncanfinney/bevy-in-app/blob/main/assets/github/profiles/120hz_fixed.png?raw=true)

**Comment by Vricken** [comment]
is this still open or did we establish that there is no real issue and the main problem is how macos display cpu usage in a weird way (where 80% is 80% of one core and not 80% of the whole cpu)

**Comment by UkoeHB** [comment]
It should not require 80% of a core to run a trivial app.

**Comment by hacknus** [comment]
> is this still open or did we establish that there is no real issue and the main problem is how macos display cpu usage in a weird way (where 80% is 80% of one core and not 80% of the whole cpu)

this is still relevant.
macOS does display the CPU usage in a different way than windows, but compared to other macOS apps, trivial bevy apps use much more CPU and it is also noticeable in battery drain.

**Comment by jason-downs** [comment]
> is this still open or did we establish that there is no real issue and the main problem is how macos display cpu usage in a weird way (where 80% is 80% of one core and not 80% of the whole cpu)

It seems to me it's real and not just a display issue; on my M4 mac the fans are always on full blast with even a relatively simple bevy app running and it eats up battery life like no tomorrow.

**Comment by phire** [comment]
> It should not require 80% of a core to run a trivial app.

It's not using 80% of a full CPU core. It's using 80% of a core that's in a pretty aggressive power saving mode. Much more aggressive than you find on any x86 core I'm aware of. Basically, "cpu usage" is now a pretty useless metric.

Despite that:

> It seems to me it's real and not just a display issue; on my M4 mac the fans are always on full blast with even a relatively simple bevy app running and it eats up battery life like no tomorrow.

Yes. To summarise my comments from further up in this thread:

Despite the the fact that "cpu usage" is now a useless metric, if you actually look at power consumption (and compare it to a roughly equivalent test app made without bevy), the bug is real. Power consumption (especially for an empty scene) is too high. 

I'm just not sure how much of this wasted resources (if any) is Apple Silicon specific, and how much is just cross-platform wastage that is hidden by the worse power management on other CPUs. I spent quite a bit of time profiling, and could never really narrow it down to a single smoking gun. Just seemed to be a "death by a thousand cuts" situation. 

For example: An empty scene has no lights. But Bevy still spends quite a bit of cpu time each frame binning the zero lights to implement clustered lighting. Doesn't take too much time to do this, but is still a small but notable overhead. There are dozens of minor inefficiencies like this which add up.

**Comment by alice-i-cecile** [comment]
I think that minor efficiencies are good to fix, but not the root cause of this particular problem :) This is *worse* on Mac than other platforms.

**Comment by phire** [comment]
> but not the root cause of this particular problem

I'd love to agree with you (and would love for someone to prove me wrong), but my several days of profiling and hacking could never find a root cause other than "many minor inefficiencies stacking together"

My personal theory is that the M1 has entered a throttling mode where the DRAM has been put into deep sleep for as long as possible, and any memory access misses cache are way expensive than normal, as the DRAM has to be bought out of sleep. Basically, a large chunk of that 50-80% of CPU utilisation is being spent with the CPU stalled for thousands of cycles while the DRAM comes out of sleep after each miss.   
If the theory is correct, the solution won't really be about reducing CPU usage, but more about reducing the amount of memory that each frame touches, trying to fit as much as possible in the L2 cache. 

> This is *worse* on Mac than other platforms.

I couldn't prove or disprove this statement.  
It's absolutely more obviously a problem on Apple Silicon, but the empty scene also used more CPU usage than I'd expect on my x86 laptop (even if it was still under 5%, I just really think 0-1% should be possible). Would be interesting to do some similar tests on non-Apple ARM SoCs. 

One potential apple-only thing my profiling ([see previous comment](https://github.com/bevyengine/bevy/issues/10261#issuecomment-3149706296)) highlighted, is that Bevy is recreating binding groups every frame (which it shouldn't), and that creating binding groups seemed to be pretty expensive on apple (which might be a wgpu deficiency). 

But that issue was nowhere near big enough to be the root cause. It barely showed up in CPU profiles (I only spotted it because it showed up as GPU memory allocation and free every frame)

**Comment by dmlary** [comment]
As someone who did deep analysis on this issue both on intel and M-series cpus, it’s a real problem. It’s a large part of why I moved on from bevy; I was tired of my legs burning when I was running my level editor on my laptop.

**Comment by ximing-xu** [comment]
this is truly bevy's bug on macos and ios!!! even simple apps on macos and ios, will trigger abnormal CPU&GPU usage，espacially on iOS，it will make phone turn to be hot in seconds. on windows and android，bevy perform well.

**Comment by Kees-van-Beilen** [comment]
Hey I'm rejoinging this thread. I could not work on bevy for some time as on my machine bevy takes ~ 100+ GB. But I cleared some storage.

I'm going do a very extensive analyses on the performance. I got some new theories, and want to try them out.

We know almost surely that the reported cpu% for mac is incorrect. This can indicate an accidental busy wait -- probably somewhere near world->render sync. I like @phire theory. I'll be looking into that one also (As Apple silicon has pretty interesting RAM).

---
Prelim result:
From a few traces, I found even more metal api misuse (one of which is inside wgpu itself, the rest is on bevy). Theses misuses make it harder to track the actual problems. For instance apple recommend ~1 command buffer be submitted per frame [1](https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/CommandBuffers.html). Bevy (and wgpu) creates >100 (some of them are empty!). ~40 of these buffers come from bevy's render pass, and ~80 from wgpu (Pree pass and transit) in one of my test with ~ 300% cpu usage This accounted for almost 100%. This can cause cpu-gpu sync issues (haven't proven this)(which are sys-time, and are thus reported in the cpu%)

I'm going to do some surgery on the bevy renderer to isolate this. But it would be nice if someone who is more familiar with the renderer and wgpu, can help come up with some architectural changes.

In one of my prev issues that I opent I also referenced that the MTLBuffers should be reused, not reallocated every frame. Those reallocs cost ~3% cpu



---

I've tried applying patch: https://github.com/liamaharon/wgpu.git and it reduced commandbuffer/frame and the cpu%

**Comment by IceSentry** [comment]
> One potential apple-only thing my profiling (https://github.com/bevyengine/bevy/issues/10261#issuecomment-3149706296) highlighted, is that Bevy is recreating binding groups every frame (which it shouldn't), and that creating binding groups seemed to be pretty expensive on apple (which might be a wgpu deficiency).

We do cache the ones that are easy to cache. Also, if you think it's an issue please submit PRs or review PRs that tries to fix these things. I've had PRs that add caching for more bind groups take months to merge because nobody was interested in reviewing them. We need community reviews to get things merged so please get involved instead of just saying we aren't doing things correctly.

**Comment by IceSentry** [comment]
Also, please keep in mind that we have spent essentially no energy optimizing empty apps. We did spend a ton of energy optimizing apps with a ton of entities being rendered. These workflows can have very different solutions. While it's obviously true that our empty or low complexity scenes performance isn't ideal please keep in mind that it's not a workflow we've tried optimizing at all.

**Comment by wentao** [comment]
I think there are four separate questions here that would be useful to distinguish:

1. **Is there an issue?**  
   Based on the reports of excess power consumption, heat, fan activity, and comparisons with equivalent non-Bevy applications, I think the answer is yes. macOS’s CPU percentage can make the magnitude look misleading, but that does not explain away the battery and thermal impact. Whether the cause is one macOS-specific bug or many small inefficiencies is a separate question.

2. **Is it worth prioritizing?**  
   That is ultimately a maintainer and project decision. Bevy may reasonably prioritize large, complex scenes over empty or low-complexity applications. However, low baseline power usage matters for editors, UI-heavy applications, laptops, and mobile devices. If these are intended Bevy use cases, this deserves consideration beyond being treated as an empty-scene microbenchmark.

3. **How should it be fixed?**  
   I don’t think we need to identify a single root cause before recognizing the issue. If profiling shows a “death by a thousand cuts,” incremental fixes—avoiding unnecessary work, allocations, bind-group creation, command buffers, or synchronization—are still valid. Individual findings can become smaller, actionable issues or PRs while this issue tracks the user-visible outcome.

4. **How should known limitations be communicated?**  
   I think the Bevy project should collect significant, repeatedly reported platform issues and publish them in a visible “known limitations” or “platform status” section on the Bevy website. That page could describe affected platforms and workloads, available workarounds, investigation status, and whether an issue is prioritized, deferred, or awaiting contributors.

GitHub issues alone are not sufficient discovery for developers evaluating whether Bevy is suitable for a project. It is a painful experience to invest in Bevy, encounter serious thermal, battery, rendering, or platform-specific problems unexpectedly, and only then discover a years-long issue thread. Some developers will understandably conclude that they cannot ship with Bevy and walk away.

I don’t expect volunteer maintainers to guarantee that every bug will be fixed or to promise timelines when no one owns the work. However, supporting a platform should involve more than compiling and launching on it. I would expect reasonable behavior for representative workloads, active triage of major reports, and transparent communication of known limitations.

It would therefore help if maintainers could state whether this behavior is:

- accepted as a real performance problem;
- considered outside Bevy’s expected macOS quality level;
- currently prioritized, deferred, or awaiting a contributor;
- covered by an agreed benchmark, preferably using frame rate and energy consumption rather than CPU percentage alone; and
- documented publicly as a known limitation until it is resolved.

This would not require promising a fix. It would give users enough information to make responsible technical decisions and contributors a concrete target for improvement.

**Comment by alice-i-cecile** [comment]
This is a real, serious performance issue. I don't have a Mac, and cannot fix or test it. PRs welcome. Energy consumption is probably a more reasonable cross-device place to look, or frame time on a fixed scene.

**Comment by Kees-van-Beilen** [comment]
There is a problem with battery consumption and the power consumption metrics. First if you calculating battery consumption it will be noisy as you can't isolate battery consumption to just your bevy app. And the power consumption metric is calculated .... based on cpu% (and a few other factors like wakeup) and I don't know if all gpu related things are a part of that metric as well. 

I recommend fixes that target bevy's rendering to look at avg Command Buffers allocated per frame, the avg MTLBuffers allocated per frame. These are directly correlated with the high cpu% reported. 

Maybe it would be an idea to create an organized group, to tackle this problem
